docs/_includes/Types.md

```diff
@@ -1511,21 +1511,131 @@ where _members_ can include fields, functions, methods and declarations of neste
+"representation") of type `set<object>`.
+
+It serves no purpose (but does no harm) to explicitly list `object` as
+an extendee in a class or trait declaration.
+
+Trait `object` contains no members.
+
+## Inheritance {#sec-inheritance}
+
+The purpose of traits is to be able to express abstraction: a trait
+encapsulates a set of behaviors; classes and traits that extend it
+_inherit_ those behaviors, perhaps specializing them.
+
+A trait or class may extend multiple other traits.
+The traits syntactically listed in a trait or class's `extends` clause
+are called its _direct parents_; the _transitive parents_ of a trait or class
+are its direct parents, the transitive parents of its direct parents, and
+the `object` trait (if it is not itself `object`).
+These are sets of traits, in that it does not matter if
+there are repetitions of a given trait in a class or trait's direct or
+transitive parents. However, if a trait with type parameters is repeated,
+it must have the same actual type parameters in each instance.
+Furthermore, a trait may not be in its own set of transitive parents; that is,
+the graph of traits connected by the directed _extends_ relationship may not
+have any cycles.
+
+A class or trait inherits (as if they are copied) all the instance members
+of its transitive parents. However, since names may not be overloaded in
+Dafny, different members (that is, members with different type signatures)
+within the set of transitive parents and the class or trait itself must have different names.[^overload]
+This restriction does mean that traits from different sources that
+coincidentally use the same name for different purposes cannot be combined
+by being part of the set of transitive parents for some new trait or class.
+
+Static members of a trait may not be redeclared and must be declared with a body.
+
+[^overload]: It is possible to conceive of a mechanism for disambiguating
+conflicting names, but this would add complexity to the language that does not
+appear to be needed, at least as yet.
+
+Where traits within an extension hierarchy do declare instance members with the same
+name (and thus the same signature), some rules apply. Recall that, for methods,
+every declaration includes a specification; if no specification is given
+explicitly, a default specification applies. Instance method declarations in traits,
+however, need not have a body.
+
+For a given non-static method M,
+
+* A trait or class may not redeclare M if it has a transitive parent that declares M and provides a body.
+* A trait may but need not provide a body if all its transitive parents that declare M do not declare a body.
+* A trait or class may not have more than one transitive parent that declares M with a body.
+* A class that has one or more transitive parents that declare M without a body
+and no transitive parent that declares M with a body must itself redeclare M
+with a body.
+* Currently (and under debate), the following restriction applies:
```

Comment:
I'm thinking that this bullet and the ones above would benefit from two definitions (_override_: noun, and _overrides_: transitive verb):

* A member `M` declared in a class or trait is an _override_ if one or more of the transitive parents also declares an `M`. If a member is not an override, it is called an _original_ member.
* Consider an override `M` in a class or trait `C` where exactly one transitive parent, say `P`, declares `M`. The member `C.M` is said to _override_ the original `P.M`.
* Consider a class or trait `C` where more than one transitive parent declares a member `M`. For any two distinct transitive parents `P` and `Q` of `C` that each declares an `M`, it must be that either `P.M` overrides `Q.M` or `Q.M` overrides `P.M`. Consequently, among all the transitive parents of `C`, the overrides of a member `M` can be linearly ordered down to the original `M`, and there is a declaration of `M` that has no overrides.
* Consider an override `M` in a class or trait `C` where more than one transitive parent declares `M`. The member `C.M` _overrides_ the `M` in the transitive parents that among the transitive parents has no override.

Reply:
Revised and simplified the text